src/util/amount.rs

```diff
@@ -0,0 +1,901 @@
+                0
+            } else {
+                precision_diff
+            }
+        };
+
+        let mut decimals = None;
+        let mut value: u64 = 0; // as satoshis
+        for c in s.chars() {
+            match c {
+                '0'...'9' => {
+                    // Do `value = 10 * value + digit`, catching overflows.
+                    match 10_u64.checked_mul(value) {
+                        None => return Err(ParseAmountError::TooBig),
+                        Some(val) => match val.checked_add((c as u8 - b'0') as u64) {
+                            None => return Err(ParseAmountError::TooBig),
+                            Some(val) => value = val,
+                        },
+                    }
+                    // Increment the decimal digit counter if past decimal.
+                    decimals = match decimals {
+                        None => None,
+                        Some(d) if d < max_decimals => Some(d + 1),
+                        _ => return Err(ParseAmountError::TooPrecise),
+                    };
+                }
+                '.' => match decimals {
+                    None => decimals = Some(0),
+                    // Double decimal dot.
+                    _ => return Err(ParseAmountError::InvalidFormat),
+                },
+                _ => return Err(ParseAmountError::InvalidFormat),
+            }
+        }
+
+        // Decimally shift left by `max_decimals - decimals`.
+        let scale_factor = max_decimals - decimals.unwrap_or(0);
+        for _ in 0..scale_factor {
+            value = match 10_u64.checked_mul(value) {
+                Some(v) => v,
+                None => return Err(ParseAmountError::TooBig),
+            };
+        }
+
+        Ok(Amount::from_sat(value))
+    }
+
+    /// Express this Amount as a floating-point value in the given denomination.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn as_float_denom(&self, denom: Denomination) -> f64 {
+        (self.as_sat() as f64) * 10_f64.powi(denom.precision())
+    }
+
+    /// Express this Amount as a floating-point value in Bitcoin.
+    ///
+    /// Equivalent to `as_float_denom(Denomination::Bitcoin)`.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn as_btc(&self) -> f64 {
+        self.as_float_denom(Denomination::Bitcoin)
+    }
+
+    /// Convert an amount in floating-point notation with a given denomination.
+    /// Can return [ParseAmountError::TooBig] or [ParseAmountError::TooPrecise].
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn from_float_denom(value: f64, denom: Denomination) -> Result<Amount, ParseAmountError> {
+        if value < 0.0 {
+            return Err(ParseAmountError::Negative);
+        }
+        if value > Amount::max_value().as_float_denom(denom) {
+            return Err(ParseAmountError::TooBig);
```

Comment:
I think there is something wrong here because the following works:
```
assert_eq!(f(18446744073709551617.0, D::Satoshi), Ok(Amount(0)));
```

Reply:
Yeah, I've noticed this. At the u64::max_value level, f64 doesn't have enough precision to distinguish a single satoshi :/

I think instead of using `u64::max_value()`, I'll look up what is the highest number f64 can represent with at least 1.0 precision. And then error if it's bigger than that.

I actually re-implemented as_btc parsing to work without floats and added some more edge-case tests. Let me port over those tests.

Reply:
I added you test case and it returns the correct error now.